File: streamlit/whoop/ots.py
import numpy as np
import pandas as pd


def _slope(y_values: np.ndarray) -> float:
    return np.polyfit(np.arange(len(y_values)), y_values, 1)[0]


def calculate_overtraining_score(recovery_df: pd.DataFrame, cycle_df: pd.DataFrame) -> dict | None:
    if recovery_df.empty or cycle_df.empty:
        return None

    merged = pd.merge(
        recovery_df[["date", "hrv", "rhr", "recovery"]],
        cycle_df[["date", "strain"]],
        on="date",
        how="inner",
    )
    merged = merged.dropna(subset=["hrv", "rhr", "recovery", "strain"])

    if len(merged) < 7:
        return None

    window = merged.tail(7).reset_index(drop=True)

    hrv_slope = _slope(window["hrv"].values)
    rhr_slope = _slope(window["rhr"].values)
    rec_slope = _slope(window["recovery"].values)
    strain_slope = _slope(window["strain"].values)

    hrv_signal = 1 if hrv_slope < 0 else 0
    rhr_signal = 1 if rhr_slope > 0 else 0
    rec_signal = 1 if rec_slope < 0 else 0
    strain_elevated = strain_slope >= -0.1

    ots_score = (hrv_signal + rhr_signal + rec_signal) * int(strain_elevated)

    if ots_score == 0:
        level = "low"
        label = "No overtraining signals detected"
        color = "#00d4aa"
    elif ots_score <= 2:
        level = "moderate"
        label = f"Moderate OTS risk — {ots_score} of 3 stress signals present"
        color = "#ffaa00"
    else:
        level = "high"
        label = "High OTS risk — all stress signals firing with sustained strain"
        color = "#ff6b6b"

    return {
        "score": ots_score,
        "level": level,
        "label": label,
        "color": color,
        "slopes": {
            "hrv": round(hrv_slope, 3),
            "rhr": round(rhr_slope, 3),
            "recovery": round(rec_slope, 3),
            "strain": round(strain_slope, 3),
        },
        "signals": {
            "hrv": bool(hrv_signal),
            "rhr": bool(rhr_signal),
            "recovery": bool(rec_signal),
            "strain_elevated": strain_elevated,
        },
        "window": window,
    }

File: apps/web/src/lib/analytics/ots.ts
import { linearSlope } from "@/lib/stats";

export type OTSResult = {
  score: 0 | 1 | 2 | 3;
  level: "low" | "moderate" | "high";
  label: string;
  color: string;
  slopes: { hrv: number; rhr: number; recovery: number; strain: number };
  signals: { hrv: boolean; rhr: boolean; recovery: boolean; strainElevated: boolean };
};

type RecoveryInput = {
  date: string;
  hrv: number | null;
  rhr: number | null;
  recovery_score: number | null;
};

type CycleInput = {
  date: string;
  strain: number | null;
};

type Joined = { date: string; hrv: number; rhr: number; recovery: number; strain: number };

export function computeOTS(recovery: RecoveryInput[], cycles: CycleInput[]): OTSResult | null {
  if (recovery.length === 0 || cycles.length === 0) return null;

  const strainByDate = new Map<string, number>();
  for (const c of cycles) {
    if (c.strain != null && Number.isFinite(c.strain)) strainByDate.set(c.date, c.strain);
  }

  const joined: Joined[] = [];
  for (const r of recovery) {
    const strain = strainByDate.get(r.date);
    if (
      strain == null ||
      r.hrv == null || !Number.isFinite(r.hrv) ||
      r.rhr == null || !Number.isFinite(r.rhr) ||
      r.recovery_score == null || !Number.isFinite(r.recovery_score)
    ) continue;
    joined.push({ date: r.date, hrv: r.hrv, rhr: r.rhr, recovery: r.recovery_score, strain });
  }

  if (joined.length < 7) return null;

  joined.sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));
  const window = joined.slice(-7);

  const hrvSlope = linearSlope(window.map((w) => w.hrv));
  const rhrSlope = linearSlope(window.map((w) => w.rhr));
  const recSlope = linearSlope(window.map((w) => w.recovery));
  const strainSlope = linearSlope(window.map((w) => w.strain));

  const hrvSignal = hrvSlope < 0;
  const rhrSignal = rhrSlope > 0;
  const recSignal = recSlope < 0;
  const strainElevated = strainSlope >= -0.1;

  const fired = (hrvSignal ? 1 : 0) + (rhrSignal ? 1 : 0) + (recSignal ? 1 : 0);
  const score = (strainElevated ? fired : 0) as 0 | 1 | 2 | 3;

  let level: OTSResult["level"];
  let label: string;
  let color: string;
  if (score === 0) {
    level = "low";
    label = "No overtraining signals detected";
    color = "#00d4aa";
  } else if (score <= 2) {
    level = "moderate";
    label = `Moderate OTS risk — ${score} of 3 stress signals present`;
    color = "#ffaa00";
  } else {
    level = "high";
    label = "High OTS risk — all stress signals firing with sustained strain";
    color = "#ff6b6b";
  }

  return {
    score,
    level,
    label,
    color,
    slopes: {
      hrv: round3(hrvSlope),
      rhr: round3(rhrSlope),
      recovery: round3(recSlope),
      strain: round3(strainSlope),
    },
    signals: {
      hrv: hrvSignal,
      rhr: rhrSignal,
      recovery: recSignal,
      strainElevated,
    },
  };
}

function round3(n: number): number {
  return Math.round(n * 1000) / 1000;
}
